ec/ecresource/deploymentresource/deployment/v2/deployment_update_payload.go

```diff
@@ -60,48 +63,63 @@ func (plan DeploymentTF) getBaseUpdatePayloads(ctx context.Context, client *api.
 		Kibana:             template.DeploymentTemplate.Resources.Kibana,
 	}
 
-	// If the deployment template has changed then we should use the template migration API
-	// to build the base update payloads
-	if newDtId != prevDtId && prevDtId != "" {
-		// Get an update request from the template migration API
-		migrateUpdateRequest, err := client.V1API.Deployments.MigrateDeploymentTemplate(
-			deployments.NewMigrateDeploymentTemplateParams().WithDeploymentID(plan.Id.ValueString()).WithTemplateID(newDtId),
-			client.AuthWriter,
-		)
+	planHasNodeTypes, diags := elasticsearchv2.PlanHasNodeTypes(ctx, plan.Elasticsearch)
+
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	templateChanged := newDtId != prevDtId && prevDtId != ""
+
+	// If the deployment template has changed or MigrateToLatestHardware is true, we should use the template migration
+	// API to build the base update payloads
+	migrateToLatest := plan.MigrateToLatestHardware.ValueBool() || templateChanged
+
+	// Template migration isn't available for deployments using node types
+	if migrateToLatest && !planHasNodeTypes {
```

Comment:
Does it make sense to make a check similar to `CheckAvailableMigration` here if template hasn't changed? Otherwise we will be calling migrate API or using cached migration request for every `Update` call if `migrateToLatest` is set - do we want this?

Reply:
I think this shouldn't be a problem because at this point we already defined that the state will be used for the fields, right? So, it shouldn't make a difference whether we use the base payloads or the payloads from the migrate request (but, please correct me if I'm wrong, as I may be confused... 🤔).

Reply:
Do you mean the logic in the plan modifiers? Is so, than I think the answer is no. E.g. the plan modifier sets the plan value for `Apm` (e.g. IC) to its current state, so the attribute is omitted from plan output. Then `Update` functions creates payload based on data from migration request that can have different values from the current state, sends it to backend and calls `read` that reads the new `Apm`'s state. If the new state of `Apm` indeed differs from the previous (that was used by the plan modifiers before `Update`), TF reports an inconsistency error in the provider.
Anyway, I think we don't want to specify payload from the migrate request if there is no migration available.

Reply:
I see, that makes sense! Thanks for the clarification 🙇‍♀️ 
Updated in https://github.com/elastic/terraform-provider-ec/pull/755/commits/35aaecb011f9d4105b8e65a10fd28ebd0473b570.